### PR TITLE
Update KMS documentation

### PR DESCRIPTION
**Pull request by JoaoSa-da**

### CODE REVIEW

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 1]
```diff
@@ -476,14 +507,56 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+with other nodes running a ``kms`` or ``jce`` providers.
+
+An example configuration that puts it all together is below:
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/encrypted-store-enabled.conf
+   :language: none
+
+.. warning::
+    Once a node starts with a KMS as its provider it can no longer be reverted without a full reset of the node
+    (i.e., database pruning and re-generation of all keys).
```
Comment (soren-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:548]: pruning is a specific term in canton, we shouldn't conflate it here

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 2]
```diff
@@ -476,14 +507,56 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+
 Auditability
 ^^^^^^^^^^^^
 
 AWS provides  tools to monitor KMS keys. To set automatic external logging, refer to the `official documentation
 <https://docs.aws.amazon.com/kms/latest/developerguide/monitoring-overview.html>`_.
 This includes instructions on how to set AWS Cloud Trail or Cloud Watch Alarms
-to keep track of usage of KMS keys. Errors resulting from the use of the
-wrapper key (i.e., during encryption and decryption) are logged in Canton.
+to keep track of usage of KMS keys. Errors resulting from the use of a
```
Comment (soren-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:558]: we also need to extend that with an optional audit trail for all the KMS usage when used in full KMS mode

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 3]
```diff
@@ -336,37 +349,93 @@ Key Management Service Setup
...
 
-KMS support can be enabled for a new installations (i.e., during the node
+KMS support can be enabled for a new installation (i.e., during the node
 bootstrap) or for an existing deployment that is transparently updated to use KMS.
-When the KMS is enabled after a node has been running, the keys are transparently
-encrypted and stored in this encrypted form in the Canton node's database.
+When the KMS is enabled after a node has been running, (a) the keys are encrypted and stored in this encrypted form
+in the Canton node's database, or (b) transparently replaced by external KMS keys.
+
+Note: If we encrypt the private keys at rest, scenario (a), then the AWS KMS keys used to encrypt them need
```
Comment (carrielaben-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:375]:
```suggestion
Note: In scenario (a), then the AWS KMS keys used to encrypt the private keys need
```

**Review by carrielaben-da — APPROVED** [review]
Some minor changes.

File: docs/2.7.0/docs/canton/architecture/security.rst [hunk 4]
```diff
@@ -119,26 +139,28 @@ The rotation frequency is fixed and cannot be changed.
...
 - The keys must not be part of Canton’s container images.
-    - The Canton private keys are stored in the database of the node and not in the container image. The wrapper key is stored in the external KMS. Credentials to access the KMS can be passed in via the environment when a container is created, the credentials must not be stored in the image.
+    - The Canton private keys are stored in the (a) database of the node or directly in the (b) external KMS and not in the container image. Credentials to access the KMS can be passed in via the environment when a container is created, the credentials must not be stored in the image.
 - A key administrator can both rotate the KMS key or long-term keys in Canton.
     - Canton already supports manual rotation of long-term keys.
-    - Early access: Support of KMS key rotation based on AWS KMS automated annual key rotation.
-    - Future support of manual KMS key rotation by re-encrypting the Canton private keys.
+    - Support of KMS key rotation based on AWS KMS automated annual key rotation.
```
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/architecture/security.rst:158]: Ok well this is something that I completely missed. I need to disable it automatic rotation or if it fails do a public operation try and update the public material...

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 5]
```diff
@@ -38,22 +38,35 @@ Symmetric Encryption:
...
+
 Key Generation and Storage
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Keys are generated in the node and stored in the node's primary storage. We currently
-support an early access version of Canton that can use an AWS key management system (KMS)
-to protect Canton's private keys. This version is available only as part of Daml Enterprise.
+Keys can either be generated in the node and stored in the node's primary storage or generated and stored by
+an external key management system (KMS).
+We currently support an early access version of Canton that can use AWS KMS to either:
```
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:56]: I removed _early access_ and created an issue #13623.

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 6]
```diff
@@ -336,37 +349,93 @@ Key Management Service Setup
...
+node's configuration file. A KMS is configured in the following way:
+
+.. _kms_config:
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/encrypted-store-enabled-tagged.conf
+   :language: none
+   :start-after: user-manual-entry-begin: KmsConfig
+   :end-before: user-manual-entry-end: KmsConfig
+
+- ``type`` specifies which KMS to use: currently only `aws` is supported.
```
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:393]: https://github.com/DACH-NY/canton/issues/13623

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 7]
```diff
@@ -336,37 +349,93 @@ Key Management Service Setup
...
+
+Specific to AWS:
+
+- ``region`` specifies which region the AWS KMS is bound to.
+- ``multi-region-key`` flag enables the replication of keys generated by the KMS. With replication turned on, the operator can replicate a key from one region to another (Note: replication of a key is not done automatically by Canton) and change the region configured in Canton at a later point in time without any other key rotation required. **The standard single-region approach is applicable for most scenarios**.
+
+Configure AWS Credentials and Permissions
```
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:400]: I added this comment to the [GCP issue](https://github.com/DACH-NY/canton/issues/13623)

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 8]
```diff
@@ -476,14 +507,56 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+
+.. warning::
+    You cannot mix an external private key storage configuration
+    with an encrypted private key storage configuration.
+
+In the example below we configure a Canton participant node (called ``participant1``) to generate and
+store private keys in an external KMS. Besides the previously presented :ref:`AWS KMS configuration <kms_config>`
+you only need to specify the correct crypto provider ``kms``:
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/encrypted-store-enabled-tagged.conf
```
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:521]: I know. I made a note to myself to include this in the integration test PR to tag in here.

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 9]
```diff
@@ -476,14 +507,56 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+Signing:
+
+- ECDSA with P-256 and P-384
+
+Asymmetric Encryption:
+
+- RSA 2048 with OAEP using SHA-256
+
+Therefore, a node running with a ``kms`` provider is only ever able to communicate
+with other nodes running a ``kms`` or ``jce`` providers.
```
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:539]: Is the domain config going to be any different no? We just need it to also be _jce_ or _kms_ no?

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 10]
```diff
@@ -476,14 +507,56 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+
+An example configuration that puts it all together is below:
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/encrypted-store-enabled.conf
```
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:543]: No it's not (it's on purpose). We need to merge the snippet in Canton before I can tag it here.

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 11]
```diff
@@ -476,14 +507,56 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+   :language: none
+
+.. warning::
+    Once a node starts with a KMS as its provider it can no longer be reverted without a full reset of the node
```
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:547]: 1) We cannot export the private keys in KMS this is not allowed (if you are the manager of the store)... 2) could potentially but this requires a substantial refactoring (don't think it's going to be in 2.7). Therefore, we should have a warning on this even if it's temporary no?

File: docs/2.7.0/docs/canton/usermanual/security.rst (hunk 2 above)
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:558]: _...AWS Cloud Trail or Cloud Watch Alarms to keep track of usage of KMS keys or of the KMS operations in general._

File: docs/2.7.0/docs/canton/usermanual/security.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (soren-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:545]: the problem is the following setup:
- domain using JCE, which has a larger set of supported algorithms than KMS and by default the required schemes on the domain are the supported algorithms of the provider
- participant is using KMS, which only supports a subset of the JCE supported algorithms, so it won't be able to connect to the domain.

the domain has to be explicitly configured to use the KMS supported algorithms as the required algorithms on the domain.

File: docs/2.7.0/docs/canton/usermanual/security.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (JoaoSa-da) [reply docs/2.7.0/docs/canton/usermanual/security.rst:553]: I don't think so... at least in AWS 
_No. All KMS keys or the private portion of an asymmetric KMS key cannot be exported in plain text from the HSMs. Only the public portion of an asymmetric KMS key can be exported from the console or by calling the GetPublicKey API._

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 12]
```diff
@@ -25,35 +25,30 @@ Supported Cryptographic Schemes in Canton
...
-Asymmetric Encryption:
-
-- ECIES on P-256 with HMAC-SHA256 and AES128-GCM (default)
-- ECIES on P-256 with HMAC-SHA256 and AES128-CBC
-- RSA 2048 with OAEP using SHA-256
-
-MAC:
-
-- HMAC with SHA-256
+.. image:: ./images/canton_supported_schemes.png
```
Comment (soren-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:28]: please use sphinx tables and not a picture (easier to change, searchable): https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#grid-tables

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 13]
```diff
@@ -476,15 +494,62 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
 KMS configuration. If you wish to change the key specification (e.g. enable multi region) you are required
 to update the configuration before rotating the wrapper key.
 
+Canton Configuration for External Key Storage and Usage
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. warning::
```
Comment (soren-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:500]: let's move this to note and at the end of this section

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 14]
```diff
@@ -476,15 +494,62 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+In the example below we configure a Canton participant node (called ``participant1``) to generate and
+store private keys in an external KMS. Besides the previously presented :ref:`AWS KMS configuration <kms_config>`
+you only need to specify the correct crypto provider ``kms``:
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/encrypted-store-enabled-tagged.conf
+   :language: none
+   :start-after: user-manual-entry-begin: PrivateKeyStoreConfig
+   :end-before: user-manual-entry-end: PrivateKeyStoreConfig
+
+AWS KMS only supports the following cryptographic schemes:
```
Comment (soren-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:513]: refer to the section with the table of supported schemes, we don't need to repeat it here

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 15]
```diff
@@ -476,15 +494,62 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+- ECDSA with P-256 and P-384
+
+Asymmetric Encryption:
+
+- RSA 2048 with OAEP using SHA-256
+
+Therefore, a node running with a ``kms`` provider is only ever able to communicate
+with other nodes running a ``kms`` or ``jce`` providers. Furthermore,
+the domain has to be explicitly configured to use the KMS supported algorithms
+as the required algorithms on the domain
```
Comment (soren-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:528]:
```suggestion
as the required algorithms on the domain.
```

make sure the snippet below will also include the domain config

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 16]
```diff
@@ -476,15 +494,62 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/encrypted-store-enabled.conf
+   :language: none
+
+.. todo::
+      #. `Enable revert for a KMS provider <https://github.com/DACH-NY/canton/issues/13635>`_
+
+.. note::
+    Currently if a node starts with a KMS as its provider it can no longer be reverted without a full reset of the node
+    (i.e., re-generation of node identity and all keys). We plan to add this revert feature in the near future.
```
Comment (soren-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:540]:
```suggestion
    (i.e., re-generation of node identity and all keys).
```

File: docs/2.7.0/docs/canton/usermanual/security.rst [hunk 17]
```diff
@@ -476,15 +494,62 @@ You can optionally pass a wrapper key id to change to or let Canton generate a n
...
 Auditability
 ^^^^^^^^^^^^
 
 AWS provides  tools to monitor KMS keys. To set automatic external logging, refer to the `official documentation
 <https://docs.aws.amazon.com/kms/latest/developerguide/monitoring-overview.html>`_.
 This includes instructions on how to set AWS Cloud Trail or Cloud Watch Alarms
-to keep track of usage of KMS keys. Errors resulting from the use of the
-wrapper key (i.e., during encryption and decryption) are logged in Canton.
-
+to keep track of usage of KMS keys. We can also extend this with an optional audit trail for all the
```
Comment (soren-da) [inline docs/2.7.0/docs/canton/usermanual/security.rst:550]: let's document the canton audit capabilities when we have implemented them

**Pull request closed without merging**